freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/CloudIdentitySyncService.java

```diff
@@ -0,0 +1,99 @@
+
+    @Inject
+    private SdxEndpoint sdxEndpoint;
+
+    public void syncCloudIdentites(Stack stack, UmsUsersState umsUsersState, BiConsumer<String, String> warnings) {
+        if (CloudPlatform.AZURE.equalsIgnoreCase(stack.getCloudPlatform())) {
+            syncAzureObjectIds(stack, umsUsersState, warnings);
+        }
+    }
+
+    private void syncAzureObjectIds(Stack stack, UmsUsersState umsUsersState, BiConsumer<String, String> warnings) {
+        String envCrn = stack.getEnvironmentCrn();
+        LOGGER.info("Syncing Azure Object IDs for environment {}", envCrn);
+
+        Map<String, List<CloudIdentity>> userCloudIdentites = getUserCloudIdentitiesToSync(umsUsersState);
+
+        Map<String, String> userToAzureObjectIdMap = getAzureObjectIdMap(userCloudIdentites);
+        Map<String, String> groupToAzureObjectIdMap = getAzureObjectIdMap(umsUsersState.getGroupToCloudIdentityMap());
+
+        SetRangerCloudIdentityMappingRequest setRangerCloudIdentityMappingRequest = new SetRangerCloudIdentityMappingRequest();
+        setRangerCloudIdentityMappingRequest.setAzureUserMapping(userToAzureObjectIdMap);
+        setRangerCloudIdentityMappingRequest.setAzureGroupMapping(groupToAzureObjectIdMap);
+        try {
+            // TODO The SDX endpoint currently sets the config and triggers refresh. The SDX endpoint should also be updated
+            //      to allow polling the status of the refresh.
+            sdxEndpoint.setRangerCloudIdentityMapping(envCrn, setRangerCloudIdentityMappingRequest);
+        } catch (Exception e) {
+            LOGGER.warn("Failed to set cloud identity mapping for environment {}", envCrn, e);
+            warnings.accept(envCrn, "Failed to set cloud identity mapping");
+        }
+    }
+
+    private Map<String, String> getAzureObjectIdMap(Map<String, List<CloudIdentity>> cloudIdentityMapping) {
+        ImmutableMap.Builder<String, String> azureObjectIdMap = ImmutableMap.builder();
+        cloudIdentityMapping.forEach((key, cloudIdentities) -> {
+            Optional<String> azureObjectId = getOptionalAzureObjectId(cloudIdentities);
+            if (azureObjectId.isPresent()) {
+                azureObjectIdMap.put(key, azureObjectId.get());
+            }
+        });
+        return azureObjectIdMap.build();
+    }
+
+    private Optional<String> getOptionalAzureObjectId(List<CloudIdentity> cloudIdentities) {
+        List<CloudIdentity> azureCloudIdentities = cloudIdentities.stream()
+                .filter(cloudIdentity -> cloudIdentity.getCloudIdentityDomain().hasAzureCloudIdentityDomain())
+                .collect(Collectors.toList());
+        if (azureCloudIdentities.isEmpty()) {
+            return Optional.empty();
+        } else if (azureCloudIdentities.size() > 1) {
+            throw new IllegalStateException(String.format("List contains multiple azure cloud identities = %s", cloudIdentities));
```

Comment:
shouldn't we handle this exception somewhere? or is it ok if `UserSyncService#internalSynchronizeStack` handles it? 
cc @handavid

Reply:
good point. I think we can expand the try block up on line 50 to start on line 42 instead. That will catch this exception and register a warning.

Reply:
I don't think we need to fail super cleanly in this situation. This is a very unexpected scenario.

Reply:
That said, I've taken David's advice.